### ISSUE TITLE
Do not multi-map rRNA/contaminant reads

### ISSUE THREAD

**Issue by ewallace** [riboviz/RiboViz#127]
Review and change the alignment parameters for rRNA/ contaminant reads to avoid unnecessary multi-mapping.

Work on #108 showed a higher count of output reads than input reads for hisat2 alignment to rRNA. That is likely caused by multi-mapping to redundant sequences in the `rRNA.fa` file. We don't need to do this, because any read that maps even once to `rRNA.fa` should be discarded.

Aim to fix by changing the hisat2 flag *for rRNA alignment only* to report only the first alignment of a read.

**Comment by mikej888** [comment]
https://ccb.jhu.edu/software/hisat2/manual.shtml#default-mode-search-for-one-or-more-alignments-report-each comments that

> HISAT2 does not "find" alignments in any specific order, so for reads that have more than N distinct, valid alignments, HISAT2 does not guarantee that the N alignments reported are the best possible in terms of alignment score. Still, this mode can be effective and fast in situations where the user cares more about whether a read aligns (or aligns a certain number of times) than where exactly it originated.

`hisat2` has a `-k` flag: `-k <int> (default: 5) report up to <int> alns per read`

@ewallace confirmed this is the flag to use.

**Comment by ewallace** [comment]
Importantly, `-k 1` to find only one alignment.

**Comment by ewallace** [comment]
There's some weird hisat2 behaviour where the number of *unaligned* reads changes when we change to -k 1. According to my mental model, changing the number of alignments to look for should not change the number of aligned reads.

Some discussion at 14/02/20 developer meeting about whether this is due to splicing options, e.g. `--no-splice-alignment` or `--no-temp-splicesite`
https://github.com/DaehwanKimLab/hisat2/issues/222

Note also that `--max-seeds` has a default value that depends on `k`, **and** that the behaviour is qualitatively different if `-k 1`:
[For RNA-read alignment, HISAT2 skips extending seeds and reports no alignments if the number of seeds is larger than the number specified with the option](http://daehwankimlab.github.io/hisat2/manual/#reporting-options).

This would explain the discrepancy, although I don't understand the rationale for this choice on the part of hisat2.

@mikej888 points out that the answer does not affect the vignette results downstream from *_trim.fq*, so we decided this is safe to approve, and close the issue.

**Comment by mikej888** [comment]
Sample outputs:

```
$ python count_fastq.py vignette/tmp/WTnone/trim.fq 
952343
```

Default k (5):

```
$ hisat2 -p 1 -N 1 --un nonrRNA.fq -x vignette/index/yeast_rRNA -S rRNA_map.sam -U vignette/tmp/WTnone/trim.fq
952343 reads; of these:
  952343 (100.00%) were unpaired; of these:
    466464 (48.98%) aligned 0 times
    10106 (1.06%) aligned exactly 1 time
    475773 (49.96%) aligned >1 times
51.02% overall alignment rate
$ python count_bam_sam.py rRNA_map.sam
(1430213, 485879) # (sequences, mapped (primary aligned) sequences)
$ python count_fastq.py nonrRNA.fq 
466464
```

Default k (5) and `--no-temp-splicesite`:

* As above.

Default k (5) and `--no-spliced-alignment`:

```
$ hisat2 -p 1 -N 1 --no-spliced-alignment --un nonrRNA.fq -x vignette/index/yeast_rRNA -S rRNA_map.sam -U vignette/tmp/WTnone/trim.fq
952343 reads; of these:
  952343 (100.00%) were unpaired; of these:
    466474 (48.98%) aligned 0 times      # 10 more than above
    10106 (1.06%) aligned exactly 1 time
    475763 (49.96%) aligned >1 times     # 10 less than above
51.02% overall alignment rate
$ python count_bam_sam.py rRNA_map.sam 
(1430203, 485869)                        # 10 less than above
$ python count_fastq.py nonrRNA.fq 
466474                                   # 10 more than above
```

`-k 1`:

```
$ hisat2 -p 1 -N 1 -k 1 --un nonrRNA.fq -x vignette/index/yeast_rRNA -S rRNA_map.sam -U vignette/tmp/WTnone/trim.fq
952343 reads; of these:
  952343 (100.00%) were unpaired; of these:
    467194 (49.06%) aligned 0 times
    485149 (50.94%) aligned exactly 1 time
    0 (0.00%) aligned >1 times
50.94% overall alignment rate
$ python count_bam_sam.py rRNA_map.sam 
(952343, 485149) # (sequences, mapped (primary aligned) sequences)
$ python count_fastq.py nonrRNA.fq
467194
```

`-k 1` and `--no-temp-splicesite`:

* As above.

`-k 1` and `--no-spliced-alignment`:

* As above.

`-k 2`:

```
$ hisat2 -p 1 -N 1 -k 2 --un nonrRNA.fq -x vignette/index/yeast_rRNA -S rRNA_map.sam -U vignette/tmp/WTnone/trim.fq
952343 reads; of these:
  952343 (100.00%) were unpaired; of these:
    467184 (49.06%) aligned 0 times
    10106 (1.06%) aligned exactly 1 time
    475053 (49.88%) aligned >1 times
50.94% overall alignment rate
$ python count_bam_sam.py rRNA_map.sam 
(1427396, 485159) # (sequences, mapped (primary aligned) sequences)
$ python count_fastq.py nonrRNA.fq 
467184
```

`-k 2` and `--no-temp-splicesite`:

* As above.

`-k 2` and `--no-spliced-alignment`:

```
$ hisat2 -p 1 -N 1 -k 2 --no-spliced-alignment --un nonrRNA.fq -x vignette/index/yeast_rRNA -S rRNA_map.sam -U vignette/tmp/WTnone/trim.fq
952343 reads; of these:
  952343 (100.00%) were unpaired; of these:
    467194 (49.06%) aligned 0 times      # 10 more than above
    10106 (1.06%) aligned exactly 1 time
    475043 (49.88%) aligned >1 times     # 10 less than above
50.94% overall alignment rate
$ python count_bam_sam.py rRNA_map.sam 
(1427386, 485149)                        # 10 less than above
$ python count_fastq.py nonrRNA.fq       # 10 more tham above
467194                                 
```

**Comment by mikej888** [comment]
Merged via pull request #145.
